# Exploratory Analysis

How car sales work in ebay

In [15]:
import pandas as pd

In [16]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [17]:
autos

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dateCrawled</th>
      <th>name</th>
      <th>seller</th>
      <th>offerType</th>
      <th>price</th>
      <th>abtest</th>
      <th>vehicleType</th>
      <th>yearOfRegistration</th>
      <th>gearbox</th>
      <th>powerPS</th>
      <th>model</th>
      <th>odometer</th>
      <th>monthOfRegistration</th>
      <th>fuelType</th>
      <th>brand</th>
      <th>notRepairedDamage</th>
      <th>dateCreated</th>
      <th>nrOfPictures</th>
      <th>postalCode</th>
      <th>lastSeen</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>2016-03-26 17:47:46</td>
      <td>Peugeot_807_160_NAVTECH_ON_BOARD</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$5,000</td>
      <td>control</td>
      <td>bus</td>
      <td>2004</td>
      <td>manuell</td>
      <td>158</td>
      <td>andere</td>
      <td>150,000km</td>
      <td>3</td>
      <td>lpg</td>
      <td>peugeot</td>
      <td>nein</td>
      <td>2016-03-26 00:00:00</td>
      <td>0</td>
      <td>79588</td>
      <td>2016-04-06 06:45:54</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2016-04-04 13:38:56</td>
      <td>BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$8,500</td>
      <td>control</td>
      <td>limousine</td>
      <td>1997</td>
      <td>automatik</td>
      <td>286</td>
      <td>7er</td>
      <td>150,000km</td>
      <td>6</td>
      <td>benzin</td>
      <td>bmw</td>
      <td>nein</td>
      <td>2016-04-04 00:00:00</td>
      <td>0</td>
      <td>71034</td>
      <td>2016-04-06 14:45:08</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2016-03-26 18:57:24</td>
      <td>Volkswagen_Golf_1.6_United</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$8,990</td>
      <td>test</td>
      <td>limousine</td>
      <td>2009</td>
      <td>manuell</td>
      <td>102</td>
      <td>golf</td>
      <td>70,000km</td>
      <td>7</td>
      <td>benzin</td>
      <td>volkswagen</td>
      <td>nein</td>
      <td>2016-03-26 00:00:00</td>
      <td>0</td>
      <td>35394</td>
      <td>2016-04-06 20:15:37</td>
    </tr>
    <tr>
      <th>3</th>
      <td>2016-03-12 16:58:10</td>
      <td>Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$4,350</td>
      <td>control</td>
      <td>kleinwagen</td>
      <td>2007</td>
      <td>automatik</td>
      <td>71</td>
      <td>fortwo</td>
      <td>70,000km</td>
      <td>6</td>
      <td>benzin</td>
      <td>smart</td>
      <td>nein</td>
      <td>2016-03-12 00:00:00</td>
      <td>0</td>
      <td>33729</td>
      <td>2016-03-15 03:16:28</td>
    </tr>
    <tr>
      <th>4</th>
      <td>2016-04-01 14:38:50</td>
      <td>Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$1,350</td>
      <td>test</td>
      <td>kombi</td>
      <td>2003</td>
      <td>manuell</td>
      <td>0</td>
      <td>focus</td>
      <td>150,000km</td>
      <td>7</td>
      <td>benzin</td>
      <td>ford</td>
      <td>nein</td>
      <td>2016-04-01 00:00:00</td>
      <td>0</td>
      <td>39218</td>
      <td>2016-04-01 14:38:50</td>
    </tr>
    <tr>
      <th>5</th>
      <td>2016-03-21 13:47:45</td>
      <td>Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$7,900</td>
      <td>test</td>
      <td>bus</td>
      <td>2006</td>
      <td>automatik</td>
      <td>150</td>
      <td>voyager</td>
      <td>150,000km</td>
      <td>4</td>
      <td>diesel</td>
      <td>chrysler</td>
      <td>NaN</td>
      <td>2016-03-21 00:00:00<

In [18]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [22]:
autos.head()

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dateCrawled</th>
      <th>name</th>
      <th>seller</th>
      <th>offerType</th>
      <th>price</th>
      <th>abtest</th>
      <th>vehicleType</th>
      <th>yearOfRegistration</th>
      <th>gearbox</th>
      <th>powerPS</th>
      <th>model</th>
      <th>odometer</th>
      <th>monthOfRegistration</th>
      <th>fuelType</th>
      <th>brand</th>
      <th>notRepairedDamage</th>
      <th>dateCreated</th>
      <th>nrOfPictures</th>
      <th>postalCode</th>
      <th>lastSeen</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>2016-03-26 17:47:46</td>
      <td>Peugeot_807_160_NAVTECH_ON_BOARD</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$5,000</td>
      <td>control</td>
      <td>bus</td>
      <td>2004</td>
      <td>manuell</td>
      <td>158</td>
      <td>andere</td>
      <td>150,000km</td>
      <td>3</td>
      <td>lpg</td>
      <td>peugeot</td>
      <td>nein</td>
      <td>2016-03-26 00:00:00</td>
      <td>0</td>
      <td>79588</td>
      <td>2016-04-06 06:45:54</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2016-04-04 13:38:56</td>
      <td>BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$8,500</td>
      <td>control</td>
      <td>limousine</td>
      <td>1997</td>
      <td>automatik</td>
      <td>286</td>
      <td>7er</td>
      <td>150,000km</td>
      <td>6</td>
      <td>benzin</td>
      <td>bmw</td>
      <td>nein</td>
      <td>2016-04-04 00:00:00</td>
      <td>0</td>
      <td>71034</td>
      <td>2016-04-06 14:45:08</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2016-03-26 18:57:24</td>
      <td>Volkswagen_Golf_1.6_United</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$8,990</td>
      <td>test</td>
      <td>limousine</td>
      <td>2009</td>
      <td>manuell</td>
      <td>102</td>
      <td>golf</td>
      <td>70,000km</td>
      <td>7</td>
      <td>benzin</td>
      <td>volkswagen</td>
      <td>nein</td>
      <td>2016-03-26 00:00:00</td>
      <td>0</td>
      <td>35394</td>
      <td>2016-04-06 20:15:37</td>
    </tr>
    <tr>
      <th>3</th>
      <td>2016-03-12 16:58:10</td>
      <td>Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$4,350</td>
      <td>control</td>
      <td>kleinwagen</td>
      <td>2007</td>
      <td>automatik</td>
      <td>71</td>
      <td>fortwo</td>
      <td>70,000km</td>
      <td>6</td>
      <td>benzin</td>
      <td>smart</td>
      <td>nein</td>
      <td>2016-03-12 00:00:00</td>
      <td>0</td>
      <td>33729</td>
      <td>2016-03-15 03:16:28</td>
    </tr>
    <tr>
      <th>4</th>
      <td>2016-04-01 14:38:50</td>
      <td>Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$1,350</td>
      <td>test</td>
      <td>kombi</td>
      <td>2003</td>
      <td>manuell</td>
      <td>0</td>
      <td>focus</td>
      <td>150,000km</td>
      <td>7</td>
      <td>benzin</td>
      <td>ford</td>
      <td>nein</td>
      <td>2016-04-01 00:00:00</td>
      <td>0</td>
      <td>39218</td>
      <td>2016-04-01 14:38:50</td>
    </tr>
  </tbody>
</table>
</div>

From this point, the created dataframe is a german list of cars sold in eBay, the 'name' column is an aggregatted set of relevant info hard to extract and make conclusions. Most of columns are objects and all of them have relevant not null data.

In [23]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [24]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_p_s', 'model',
       'odometer', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [25]:
autos.head()

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>date_crawled</th>
      <th>name</th>
      <th>seller</th>
      <th>offer_type</th>
      <th>price</th>
      <th>abtest</th>
      <th>vehicle_type</th>
      <th>year_of_registration</th>
      <th>gearbox</th>
      <th>power_p_s</th>
      <th>model</th>
      <th>odometer</th>
      <th>month_of_registration</th>
      <th>fuel_type</th>
      <th>brand</th>
      <th>not_repaired_damage</th>
      <th>ad_created</th>
      <th>nr_of_pictures</th>
      <th>postal_code</th>
      <th>last_seen</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>2016-03-26 17:47:46</td>
      <td>Peugeot_807_160_NAVTECH_ON_BOARD</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$5,000</td>
      <td>control</td>
      <td>bus</td>
      <td>2004</td>
      <td>manuell</td>
      <td>158</td>
      <td>andere</td>
      <td>150,000km</td>
      <td>3</td>
      <td>lpg</td>
      <td>peugeot</td>
      <td>nein</td>
      <td>2016-03-26 00:00:00</td>
      <td>0</td>
      <td>79588</td>
      <td>2016-04-06 06:45:54</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2016-04-04 13:38:56</td>
      <td>BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$8,500</td>
      <td>control</td>
      <td>limousine</td>
      <td>1997</td>
      <td>automatik</td>
      <td>286</td>
      <td>7er</td>
      <td>150,000km</td>
      <td>6</td>
      <td>benzin</td>
      <td>bmw</td>
      <td>nein</td>
      <td>2016-04-04 00:00:00</td>
      <td>0</td>
      <td>71034</td>
      <td>2016-04-06 14:45:08</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2016-03-26 18:57:24</td>
      <td>Volkswagen_Golf_1.6_United</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$8,990</td>
      <td>test</td>
      <td>limousine</td>
      <td>2009</td>
      <td>manuell</td>
      <td>102</td>
      <td>golf</td>
      <td>70,000km</td>
      <td>7</td>
      <td>benzin</td>
      <td>volkswagen</td>
      <td>nein</td>
      <td>2016-03-26 00:00:00</td>
      <td>0</td>
      <td>35394</td>
      <td>2016-04-06 20:15:37</td>
    </tr>
    <tr>
      <th>3</th>
      <td>2016-03-12 16:58:10</td>
      <td>Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$4,350</td>
      <td>control</td>
      <td>kleinwagen</td>
      <td>2007</td>
      <td>automatik</td>
      <td>71</td>
      <td>fortwo</td>
      <td>70,000km</td>
      <td>6</td>
      <td>benzin</td>
      <td>smart</td>
      <td>nein</td>
      <td>2016-03-12 00:00:00</td>
      <td>0</td>
      <td>33729</td>
      <td>2016-03-15 03:16:28</td>
    </tr>
    <tr>
      <th>4</th>
      <td>2016-04-01 14:38:50</td>
      <td>Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...</td>
      <td>privat</td>
      <td>Angebot</td>
      <td>$1,350</td>
      <td>test</td>
      <td>kombi</td>
      <td>2003</td>
      <td>manuell</td>
      <td>0</td>
      <td>focus</td>
      <td>150,000km</td>
      <td>7</td>
      <td>benzin</td>
      <td>ford</td>
      <td>nein</td>
      <td>2016-04-01 00:00:00</td>
      <td>0</td>
      <td>39218</td>
      <td>2016-04-01 14:38:50</td>
    </tr>
  </tbody>
</table>
</div>

From now on, the columns have snakecase names, that easier to read and deal with Python code.

In [26]:
autos.describe()

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>year_of_registration</th>
      <th>power_p_s</th>
      <th>month_of_registration</th>
      <th>nr_of_pictures</th>
      <th>postal_code</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>50000.000000</td>
      <td>50000.000000</td>
      <td>50000.000000</td>
      <td>50000.0</td>
      <td>50000.000000</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>2005.073280</td>
      <td>116.355920</td>
      <td>5.723360</td>
      <td>0.0</td>
      <td>50813.627300</td>
    </tr>
    <tr>
      <th>std</th>
      <td>105.712813</td>
      <td>209.216627</td>
      <td>3.711984</td>
      <td>0.0</td>
      <td>25779.747957</td>
    </tr>
    <tr>
      <th>min</th>
      <td>1000.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>1067.000000</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>1999.000000</td>
      <td>70.000000</td>
      <td>3.000000</td>
      <td>0.0</td>
      <td>30451.000000</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>2003.000000</td>
      <td>105.000000</td>
      <td>6.000000</td>
      <td>0.0</td>
      <td>49577.000000</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>2008.000000</td>
      <td>150.000000</td>
      <td>9.000000</td>
      <td>0.0</td>
      <td>71540.000000</td>
    </tr>
    <tr>
      <th>max</th>
      <td>9999.000000</td>
      <td>17700.000000</td>
      <td>12.000000</td>
      <td>0.0</td>
      <td>99998.000000</td>
    </tr>
  </tbody>
</table>
</div>

nr_of_pictures column have all columns with 0 value
power_p_s and year_of_registration have weird max numbers
price and odometer should be numbers instead of strings

In [27]:
autos['price'] = (autos['price']
                  .str.replace(r'[$|,]', '')
                  .astype(float)
                 )

In [28]:
autos['odometer'] = (autos['odometer']
                     .str.replace(r'[km|,]', '')
                     .astype(int)
                    )
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

## Removing Outliers

In [29]:
autos['price'].unique().shape

(2357,)

In [30]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [31]:
outliers = autos['price'].value_counts().sort_index(ascending=True)[50:180]
outliers

50.0      49
55.0       2
59.0       1
60.0       9
65.0       5
66.0       1
70.0      10
75.0       5
79.0       1
80.0      15
89.0       1
90.0       5
99.0      19
100.0    134
110.0      3
111.0      2
115.0      2
117.0      1
120.0     39
122.0      1
125.0      8
129.0      1
130.0     15
135.0      1
139.0      1
140.0      9
145.0      2
149.0      7
150.0    224
156.0      2
160.0      8
170.0      7
173.0      1
175.0     12
179.0      1
180.0     35
Name: price, dtype: int64

Any value below 100 will be discarted. 
These values do not represent any viable car's price
100 is also the first number with more than 100 occurences

In [32]:
outliers = autos['price'].value_counts().sort_index(ascending=False).head(15)
outliers

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
Name: price, dtype: int64

Any value above 350000 will be discarted. These values are at least 2,5 times bigger than any  other in list.

In [33]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

There's no outliers in odometer_km column

In [34]:
autos = autos.loc[autos['price'].between(100, 350000)]
autos['price'].describe()

count     48224.000000
mean       5930.371433
std        9078.372762
min         100.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price, dtype: float64

In [36]:
autos.describe()

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>price</th>
      <th>year_of_registration</th>
      <th>power_p_s</th>
      <th>odometer_km</th>
      <th>month_of_registration</th>
      <th>nr_of_pictures</th>
      <th>postal_code</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>48224.000000</td>
      <td>48224.000000</td>
      <td>48224.000000</td>
      <td>48224.000000</td>
      <td>48224.000000</td>
      <td>48224.0</td>
      <td>48224.000000</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>5930.371433</td>
      <td>2004.730964</td>
      <td>117.677609</td>
      <td>125919.148142</td>
      <td>5.801634</td>
      <td>0.0</td>
      <td>50987.919729</td>
    </tr>
    <tr>
      <th>std</th>
      <td>9078.372762</td>
      <td>87.897388</td>
      <td>201.206304</td>
      <td>39543.339640</td>
      <td>3.676976</td>
      <td>0.0</td>
      <td>25737.119986</td>
    </tr>
    <tr>
      <th>min</th>
      <td>100.000000</td>
      <td>1000.000000</td>
      <td>0.000000</td>
      <td>5000.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>1067.000000</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>1250.000000</td>
      <td>1999.000000</td>
      <td>73.000000</td>
      <td>125000.000000</td>
      <td>3.000000</td>
      <td>0.0</td>
      <td>30823.000000</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>3000.000000</td>
      <td>2004.000000</td>
      <td>107.000000</td>
      <td>150000.000000</td>
      <td>6.000000</td>
      <td>0.0</td>
      <td>49716.000000</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>7499.000000</td>
      <td>2008.000000</td>
      <td>150.000000</td>
      <td>150000.000000</td>
      <td>9.000000</td>
      <td>0.0</td>
      <td>71666.750000</td>
    </tr>
    <tr>
      <th>max</th>
      <td>350000.000000</td>
      <td>9999.000000</td>
      <td>17700.000000</td>
      <td>150000.000000</td>
      <td>12.000000</td>
      <td>0.0</td>
      <td>99998.000000</td>
    </tr>
  </tbody>
</table>
</div>

Assuming that in average each car runs 10.000km per year, it is possible to infer year_of_registration for the invalid cells.

## Treating Dates

In [37]:
autos[['date_crawled', 'ad_created', 'last_seen']].describe()

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>date_crawled</th>
      <th>ad_created</th>
      <th>last_seen</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>48224</td>
      <td>48224</td>
      <td>48224</td>
    </tr>
    <tr>
      <th>unique</th>
      <td>46571</td>
      <td>76</td>
      <td>38232</td>
    </tr>
    <tr>
      <th>top</th>
      <td>2016-03-16 21:50:53</td>
      <td>2016-04-03 00:00:00</td>
      <td>2016-04-07 06:17:27</td>
    </tr>
    <tr>
      <th>freq</th>
      <td>3</td>
      <td>1874</td>
      <td>8</td>
    </tr>
  </tbody>
</table>
</div>

In [38]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025361
2016-03-06    0.014039
2016-03-07    0.036061
2016-03-08    0.033179
2016-03-09    0.033013
2016-03-10    0.032287
2016-03-11    0.032598
2016-03-12    0.036911
2016-03-13    0.015677
2016-03-14    0.036662
2016-03-15    0.034319
2016-03-16    0.029467
2016-03-17    0.031499
2016-03-18    0.012898
2016-03-19    0.034734
2016-03-20    0.037803
2016-03-21    0.037201
2016-03-22    0.032888
2016-03-23    0.032287
2016-03-24    0.029446
2016-03-25    0.031499
2016-03-26    0.032308
2016-03-27    0.031126
2016-03-28    0.034962
2016-03-29    0.034112
2016-03-30    0.033738
2016-03-31    0.031851
2016-04-01    0.033697
2016-04-02    0.035605
2016-04-03    0.038611
2016-04-04    0.036538
2016-04-05    0.013064
2016-04-06    0.003173
2016-04-07    0.001389
Name: date_crawled, dtype: float64

In [39]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033096
2016-03-10    0.031997
2016-03-11    0.032909
2016-03-12    0.036745
2016-03-13    0.017045
2016-03-14    0.035294
2016-03-15    0.034049
2016-03-16    0.029964
2016-03-17    0.031167
2016-03-18    0.013582
2016-03-19    0.033614
2016-03-20    0.037865
2016-03-21 

After the crawler implementation, ad_created registration grew 20 times.

In [40]:
autos['last_seen'].str[:10].value_counts(dropna=False).sort_index()

2016-03-05       52
2016-03-06      208
2016-03-07      262
2016-03-08      353
2016-03-09      462
2016-03-10      513
2016-03-11      598
2016-03-12     1147
2016-03-13      428
2016-03-14      609
2016-03-15      765
2016-03-16      793
2016-03-17     1355
2016-03-18      353
2016-03-19      760
2016-03-20      996
2016-03-21      991
2016-03-22     1030
2016-03-23      896
2016-03-24      953
2016-03-25      921
2016-03-26      804
2016-03-27      750
2016-03-28     1005
2016-03-29     1075
2016-03-30     1191
2016-03-31     1149
2016-04-01     1102
2016-04-02     1200
2016-04-03     1212
2016-04-04     1183
2016-04-05     6031
2016-04-06    10704
2016-04-07     6373
Name: last_seen, dtype: int64

The dates between '2016-04-05' and '2016-04-07' had significant raise. Eighter the crawler and ad creation had the same change.

In [41]:
autos['year_of_registration'].describe()

count    48224.000000
mean      2004.730964
std         87.897388
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: year_of_registration, dtype: float64

The year_of_registration column have some inconsistent values that have to be cleaned like year 1000 and 9999

In [42]:
autos['year_of_registration'].value_counts(normalize=True, dropna=False).sort_index()

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000041
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000041
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000041
1943    0.000021
1948    0.000021
1950    0.000021
1951    0.000041
1952    0.000021
1953    0.000021
1954    0.000041
1955    0.000041
1956    0.000083
1957    0.000041
1958    0.000083
1959    0.000124
1960    0.000456
1961    0.000124
1962    0.000083
1963    0.000166
1964    0.000249
          ...   
2000    0.064366
2001    0.054516
2002    0.051364
2003    0.055844
2004    0.055968
2005    0.060364
2006    0.055325
2007    0.047134
2008    0.045828
2009    0.043132
2010    0.032909
2011    0.033552
2012    0.027123
2013    0.016610
2014    0.013728
2015    0.007880
2016    0.024925
2017    0.028679
2018    0.009705
2019    0.000041
2800    0.000021
4100    0.000021
4500    0.000021
4800    0.000021
5000    0.000062
5911    0.000021
6200    0.000021
8888    0.0000

In [43]:
autos.loc[~autos['year_of_registration'].between(1900, 2016), 'year_of_registration'].count()

1872

In this analysis it is considered just modern model cars.
Despite of removing wrong data, cars older than 1900 will not be considered.

Before 1900, car's architecture were different from today. Thus, for eBay car section this type of product may have different consumers, prices and specifications.

This dataframe does not contain data input after 2016, so, car's model can't be for later then 2016.

In [44]:
autos = autos.loc[autos['year_of_registration'].between(1900, 2016)]

In [45]:
autos['year_of_registration'].value_counts(normalize=True, dropna= False).sort_values(ascending=False)

2000    0.066966
2005    0.062802
1999    0.062112
2004    0.058228
2003    0.058099
2006    0.057560
2001    0.056718
2002    0.053439
1998    0.050483
2007    0.049038
2008    0.047679
2009    0.044874
1997    0.041530
2011    0.034907
2010    0.034238
1996    0.029233
2012    0.028219
2016    0.025932
1995    0.025738
2013    0.017281
2014    0.014282
1994    0.013505
1993    0.009061
2015    0.008198
1992    0.007918
1991    0.007292
1990    0.007163
1989    0.003689
1988    0.002869
1985    0.002006
          ...   
1960    0.000475
1976    0.000453
1969    0.000410
1975    0.000388
1965    0.000367
1964    0.000259
1963    0.000173
1961    0.000129
1959    0.000129
1937    0.000086
1958    0.000086
1962    0.000086
1956    0.000086
1955    0.000043
1954    0.000043
1957    0.000043
1951    0.000043
1910    0.000043
1934    0.000043
1941    0.000043
1948    0.000022
1953    0.000022
1939    0.000022
1938    0.000022
1943    0.000022
1931    0.000022
1950    0.000022
1929    0.0000

Half of the models have registration_year between 1999 - 2009.

In [46]:
eligible_brands = (autos['brand']
                   .value_counts(normalize=True, dropna=False) > 0.005
                  ).value_counts()

eligible_brands

True     27
False    13
Name: brand, dtype: int64

27 Brands have at least 5% share of the car's catalog were considered

In [47]:
brands = (autos['brand']
          .value_counts(normalize=True)
          .sort_values(ascending=False).head(27)
         ).index

In [48]:
brand_mean_prices = {}
brand_car_miles = {}

for b in brands:
    brand_mean_prices[b] = autos.loc[autos['brand'] == b, 'price'].mean()
    brand_car_miles[b] = autos.loc[autos['brand'] == b, 'odometer_km'].mean()

brand_mean_prices

{'alfa_romeo': 4100.915857605178,
 'audi': 9380.718547986076,
 'bmw': 8381.677305658899,
 'chevrolet': 6759.885931558935,
 'citroen': 3796.26267281106,
 'fiat': 2836.8736310025274,
 'ford': 3779.269076305221,
 'honda': 4119.109589041096,
 'hyundai': 5411.075431034483,
 'kia': 6018.442073170731,
 'mazda': 4129.774787535411,
 'mercedes_benz': 8672.654241071428,
 'mini': 10639.450980392157,
 'mitsubishi': 3439.10290237467,
 'nissan': 4756.659634317863,
 'opel': 3005.4960772480385,
 'peugeot': 3113.860549132948,
 'porsche': 46788.444444444445,
 'renault': 2496.070577451879,
 'seat': 4433.419621749409,
 'skoda': 6409.609724047306,
 'smart': 3596.40273556231,
 'sonstige_autos': 12784.556561085972,
 'suzuki': 4126.341818181818,
 'toyota': 5167.091062394604,
 'volkswagen': 5436.950096948668,
 'volvo': 4993.208037825059}

Cars from 'other' brands have the most highest mean price, it can represent old brands or office brands. 

Despite this, the most expensive are: Audi, Mercedes, BMW and Mini.
On the other hand, Renault, Smart, Ford, Fiat and Citroen have the lowest mean prices

In [49]:
bmp_series = pd.Series(brand_mean_prices)
bcm_series = pd.Series(brand_car_miles)

In [50]:
b_df = pd.DataFrame(data={'mean_price':bmp_series,'mean_miles':bcm_series})

In [51]:
b_df.sort_values(by='mean_miles', ascending=False)

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mean_price</th>
      <th>mean_miles</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>volvo</th>
      <td>4993.208038</td>
      <td>138581.560284</td>
    </tr>
    <tr>
      <th>bmw</th>
      <td>8381.677306</td>
      <td>132695.320149</td>
    </tr>
    <tr>
      <th>alfa_romeo</th>
      <td>4100.915858</td>
      <td>131747.572816</td>
    </tr>
    <tr>
      <th>mercedes_benz</th>
      <td>8672.654241</td>
      <td>131025.669643</td>
    </tr>
    <tr>
      <th>opel</th>
      <td>3005.496077</td>
      <td>129384.429692</td>
    </tr>
    <tr>
      <th>audi</th>
      <td>9380.718548</td>
      <td>129245.400298</td>
    </tr>
    <tr>
      <th>volkswagen</th>
      <td>5436.950097</td>
      <td>128799.877539</td>
    </tr>
    <tr>
      <th>renault</th>
      <td>2496.070577</td>
      <td>128281.393217</td>
    </tr>
    <tr>
      <th>peugeot</th>
      <td>3113.860549</td>
      <td>127127.890173</td>
    </tr>
    <tr>
      <th>mitsubishi</th>
      <td>3439.102902</td>
      <td>127097.625330</td>
    </tr>
    <tr>
      <th>mazda</th>
      <td>4129.774788</td>
      <td>124553.824363</td>
    </tr>
    <tr>
      <th>ford</th>
      <td>3779.269076</td>
      <td>124277.108434</td>
    </tr>
    <tr>
      <th>honda</th>
      <td>4119.109589</td>
      <td>122493.150685</td>
    </tr>
    <tr>
      <th>seat</th>
      <td>4433.419622</td>
      <td>121536.643026</td>
    </tr>
    <tr>
      <th>citroen</th>
      <td>3796.262673</td>
      <td>119554.531490</td>
    </tr>
    <tr>
      <th>nissan</th>
      <td>4756.659634</td>
      <td>118326.300985</td>
    </tr>
    <tr>
      <th>fiat</th>
      <td>2836.873631</td>
      <td>116950.294861</td>
    </tr>
    <tr>
      <th>toyota</th>
      <td>5167.091062</td>
      <td>115944.350759</td>
    </tr>
    <tr>
      <th>kia</th>
      <td>6018.442073</td>
      <td>112301.829268</td>
    </tr>
    <tr>
      <th>skoda</th>
      <td>6409.609724</td>
      <td>110906.701708</td>
    </tr>
    <tr>
      <th>suzuki</th>
      <td>4126.341818</td>
      <td>108163.636364</td>
    </tr>
    <tr>
      <th>hyundai</th>
      <td>5411.075431</td>
      <td>106885.775862</td>
    </tr>
    <tr>
      <th>chevrolet</th>
      <td>6759.885932</td>
      <td>100266.159696</td>
    </tr>
    <tr>
      <th>smart</th>
      <td>3596.402736</td>
      <td>99734.042553</td>
    </tr>
    <tr>
      <th>porsche</th>
      <td>46788.444444</td>
      <td>98476.702509</td>
    </tr>
    <tr>
      <th>sonstige_autos</th>
      <td>12784.556561</td>
      <td>90395.927602</td>
    </tr>
    <tr>
      <th>mini</th>
      <td>10639.450980</td>
      <td>88308.823529</td>
    </tr>
  </tbody>
</table>
</div>